File: src/train_bot/transport_api.py
from __future__ import annotations

import datetime as dt
from typing import Any, Optional, Sequence

import httpx

from .config import TransportApiSettings
from .models import CallingPoint, Departure, StationSummary


class TransportApiError(RuntimeError):
    """Raised when TransportAPI returns an error response."""


class TransportApiClient:
    """Client for TransportAPI live departures endpoint."""

    def __init__(self, settings: TransportApiSettings) -> None:
        self._settings = settings
        self._client = httpx.AsyncClient(
            base_url=settings.base_url,
            timeout=10.0,
            headers={"Accept": "application/json"},
            follow_redirects=True,
        )

    async def close(self) -> None:
        await self._client.aclose()

    async def __aenter__(self) -> "TransportApiClient":  # pragma: no cover - convenience
        return self

    async def __aexit__(self, exc_type, exc, tb) -> None:  # pragma: no cover - convenience
        await self.close()

    async def get_departures(
        self,
        origin_crs: str,
        *,
        destination_crs: Optional[str] = None,
        limit: int = 5,
        when: Optional[dt.datetime] = None,
    ) -> Sequence[Departure]:
        params = {
            "app_id": self._settings.app_id,
            "app_key": self._settings.app_key,
            "limit": str(limit),
            "station_detail": "calling_at",
        }
        if destination_crs:
            params["calling_at"] = destination_crs
        if when:
            params["date"] = when.strftime("%Y-%m-%d")
            params["time"] = when.strftime("%H:%M")

        path = f"/uk/train/station/{origin_crs}/live.json"
        response = await self._client.get(path, params=params)
        payload = await self._json_or_error(response, "requesting departures")

        departures = payload.get("departures", {}).get("all", [])
        return [self._parse_departure(item) for item in departures][:limit]

    async def search_station(self, query: str, *, limit: int = 5) -> Sequence[StationSummary]:
        params = {
            "app_id": self._settings.app_id,
            "app_key": self._settings.app_key,
            "query": query,
            "type": "train_station",
            "limit": str(limit),
        }
        response = await self._client.get("/uk/places.json", params=params)
        payload = await self._json_or_error(response, "searching for stations")

        members = payload.get("member", [])
        results: list[StationSummary] = []
        for item in members:
            code = item.get("station_code")
            name = item.get("name")
            if not code or not name:
                continue
            results.append(StationSummary(name=name, station_code=code))
            if len(results) >= limit:
                break
        return results

    async def _json_or_error(self, response: httpx.Response, action: str) -> dict[str, Any]:
        if response.status_code >= 400:
            raise TransportApiError(
                f"TransportAPI error {response.status_code} while {action}: {response.text[:200]}"
            )
        try:
            return response.json()
        except ValueError as exc:
            snippet = response.text[:200] or "<empty body>"
            content_type = response.headers.get("content-type", "unknown")
            raise TransportApiError(
                "TransportAPI returned a non-JSON response while "
                f"{action} (status {response.status_code}, content-type {content_type}): {snippet}"
            ) from exc

    @staticmethod
    def _parse_departure(data: dict[str, Any]) -> Departure:
        station_details = data.get("station_detail") or {}
        calling_points_data = station_details.get("calling_at") or data.get("calling_at") or []
        calling_points = [
            CallingPoint(
                station_name=point.get("station_name", ""),
                station_code=point.get("station_code", ""),
                aimed_arrival_time=point.get("aimed_arrival_time") or point.get("aimed_pass_time"),
                expected_arrival_time=point.get("expected_arrival_time"),
            )
            for point in calling_points_data
        ]

        destination_name = TransportApiClient._extract_destination_name(data)
        destination_code = TransportApiClient._find_destination_code(destination_name, calling_points)

        return Departure(
            service_uid=data.get("train_uid", ""),
            destination_name=destination_name,
            destination_code=destination_code,
            platform=data.get("platform"),
            aimed_departure_time=data.get("aimed_departure_time"),
            expected_departure_time=data.get("expected_departure_time") or data.get("aimed_departure_time"),
            status=data.get("status", ""),
            calling_points=calling_points,
            operator_name=data.get("operator_name") or data.get("operator"),
        )

    @staticmethod
    def _extract_destination_name(data: dict[str, Any]) -> str:
        destination = data.get("destination_name")
        if isinstance(destination, list):
            return destination[0]
        return destination or "Unknown"

    @staticmethod
    def _find_destination_code(name: str, calling_points: Sequence[CallingPoint]) -> str:
        for point in calling_points:
            if point.station_name == name:
                return point.station_code
        return ""


def create_transport_client(settings: TransportApiSettings) -> TransportApiClient:
    return TransportApiClient(settings)

File: worker/src/providers/transportApi.ts
import type { Departure, StationSummary, CallingPoint } from "../models";
import type { RailProvider } from "./types";
import { ProviderError } from "./types";

export interface TransportApiOptions {
  appId: string;
  appKey: string;
  baseUrl: string;
}

export class TransportApiProvider implements RailProvider {
  readonly name = "TransportAPI";

  constructor(private readonly options: TransportApiOptions) {}

  async searchStation(query: string, limit = 5): Promise<StationSummary[]> {
    const url = new URL("/uk/places.json", this.options.baseUrl);
    url.searchParams.set("app_id", this.options.appId);
    url.searchParams.set("app_key", this.options.appKey);
    url.searchParams.set("query", query);
    url.searchParams.set("type", "train_station");
    url.searchParams.set("limit", String(limit));

    const payload = await this.fetchJson<{ member?: Array<Record<string, unknown>> }>(url);
    const members = payload.member ?? [];
    const results: StationSummary[] = [];
    for (const item of members) {
      const code = item.station_code;
      const name = item.name;
      if (typeof code !== "string" || typeof name !== "string") {
        continue;
      }
      results.push({ name, stationCode: code });
      if (results.length >= limit) {
        break;
      }
    }
    return results;
  }

  async getDepartures(
    originCrs: string,
    {
      destinationCrs,
      limit = 5,
      when
    }: { destinationCrs?: string; limit?: number; when?: Date }
  ): Promise<Departure[]> {
    const url = new URL(`/uk/train/station/${originCrs}/live.json`, this.options.baseUrl);
    url.searchParams.set("app_id", this.options.appId);
    url.searchParams.set("app_key", this.options.appKey);
    url.searchParams.set("limit", String(limit));
    url.searchParams.set("station_detail", "calling_at");

    if (destinationCrs) {
      url.searchParams.set("calling_at", destinationCrs);
    }
    if (when) {
      url.searchParams.set("date", formatDate(when));
      url.searchParams.set("time", formatTime(when));
    }

    const payload = await this.fetchJson<any>(url);
    const departures = payload.departures?.all ?? [];

    return departures.slice(0, limit).map((item: any) => parseDeparture(item));
  }

  private async fetchJson<T>(url: URL): Promise<T> {
    const response = await fetch(url.toString(), {
      headers: { Accept: "application/json" }
    });

    const text = await response.text();
    console.info(`[transportapi] GET ${url.pathname} -> ${response.status}`);
    console.info(`[transportapi] response body: ${text.slice(0, 500)}`);
    if (response.status === 404) {
      console.info("[transportapi] No data returned (404)");
      return {} as T;
    }
    if (!response.ok) {
      throw new ProviderError(
        `TransportAPI error ${response.status}: ${text.slice(0, 200) || "<empty>"}`
      );
    }

    try {
      return JSON.parse(text) as T;
    } catch (error) {
      console.error("[transportapi] Failed to parse JSON", text.slice(0, 200));
      throw new ProviderError(`TransportAPI returned invalid JSON: ${text.slice(0, 200)}`);
    }
  }
}

function formatDate(date: Date): string {
  const year = date.getFullYear();
  const month = String(date.getMonth() + 1).padStart(2, "0");
  const day = String(date.getDate()).padStart(2, "0");
  return `${year}-${month}-${day}`;
}

function formatTime(date: Date): string {
  const hours = String(date.getHours()).padStart(2, "0");
  const minutes = String(date.getMinutes()).padStart(2, "0");
  return `${hours}:${minutes}`;
}

function parseDeparture(data: any): Departure {
  const stationDetails = data.station_detail ?? {};
  const callingPointsData = stationDetails.calling_at ?? data.calling_at ?? [];
  const callingPoints: CallingPoint[] = Array.isArray(callingPointsData)
    ? callingPointsData.map((point: any) => ({
        stationName: point.station_name ?? "",
        stationCode: point.station_code ?? "",
        aimedArrivalTime: point.aimed_arrival_time ?? point.aimed_pass_time ?? undefined,
        expectedArrivalTime: point.expected_arrival_time ?? undefined
      }))
    : [];

  const destinationName = extractDestinationName(data);
  const destinationCode = findDestinationCode(destinationName, callingPoints);

  return {
    serviceUid: data.train_uid ?? "",
    destinationName,
    destinationCode,
    platform: data.platform ?? undefined,
    aimedDepartureTime: data.aimed_departure_time ?? undefined,
    expectedDepartureTime: data.expected_departure_time ?? data.aimed_departure_time ?? undefined,
    status: data.status ?? "",
    callingPoints,
    operatorName: data.operator_name ?? data.operator ?? undefined
  };
}

function extractDestinationName(data: any): string {
  const destination = data.destination_name;
  if (Array.isArray(destination) && destination.length) {
    return destination[0] ?? "Unknown";
  }
  if (typeof destination === "string" && destination.length) {
    return destination;
  }
  return "Unknown";
}

function findDestinationCode(name: string, callingPoints: CallingPoint[]): string {
  for (const point of callingPoints) {
    if (point.stationName === name) {
      return point.stationCode;
    }
  }
  return "";
}
